File: dailyCoding/953.js
/**
 * @param {string[]} grid
 * @return {number}
 */
var regionsBySlashes = function (grid) {
    const m = new Map();
    const N = grid.length;
    for (let i = 0; i < N; i++) {
        for (let j = 0; j < N; j++) {
            let t = grid[i][j];
            if (t === "/") {
                union(m, `${i}:${j}:${0}`, `${i}:${j}:${3}`);
                union(m, `${i}:${j}:${1}`, `${i}:${j}:${2}`);
            } else if (t === "\\") {
                union(m, `${i}:${j}:${0}`, `${i}:${j}:${1}`);
                union(m, `${i}:${j}:${2}`, `${i}:${j}:${3}`);
            } else {
                union(m, `${i}:${j}:${0}`, `${i}:${j}:${1}`);
                union(m, `${i}:${j}:${2}`, `${i}:${j}:${3}`);
                union(m, `${i}:${j}:${0}`, `${i}:${j}:${2}`);
            }
        }
    }
    for (let i = 0; i < N; i++) {
        for (let j = 0; j < N; j++) {
            if (i !== N - 1) {
                union(m, `${i}:${j}:${1}`, `${i + 1}:${j}:${3}`);
            }
            if (j !== N - 1) {
                union(m, `${i}:${j}:${2}`, `${i}:${j + 1}:${0}`);
            }
        }
    }
    let res = new Set(m.values());
    return res.size;
};

// map : Map object
// key : string type key i:j:[0123]
const find = (map, key) => {
    let t = map.get(key);
    if (t === undefined) {
        map.set(key, key);
        return key;
    }
    let tt = map.get(t);
    while (tt !== t) {
        t = tt;
        tt = map.get(t);
    }
    map.set(key, t);
    return t;
};

// map : Map
// key1, key2 : string type keys
const union = (map, key1, key2) => {
    let x = find(map, key1);
    let y = find(map, key2);

    if (x === y) {
        return;
    }
    map.set(x, y);
};

let x = regionsBySlashes(["\\/\\ ", " /\\/", " \\/ ", "/ / "])
console.log(x);

File: dailyCoding/953.py
from typing import List, Tuple, Dict


class myUnion:
    def __init__(self):
        self.pd = dict()

    def find(self, p):
        self.pd.setdefault(p, p)
        if p != self.pd[p]:
            self.pd[p] = self.find(self.pd[p])
        return self.pd[p]

    def union(self, p1, p2):
        x = self.find(p1)
        y = self.find(p2)
        if x != y:
            self.pd[x] = y
            return

    def count(self):
        res = set()
        for k in self.pd.keys():
            v = self.find(k)
            res.add(v)
        return len(res)


class Solution:
    def regionsBySlashes(self, grid: List[str]) -> int:
        N = len(grid)
        d = myUnion()
        for i, row in enumerate(grid):
            for j, v in enumerate(row):
                for t in range(4):
                    d.find((i, j, t))

                if v == '/':
                    d.union((i, j, 0), (i, j, 3))
                    d.union((i, j, 1), (i, j, 2))
                elif v == '\\':
                    d.union((i, j, 0), (i, j, 1))
                    d.union((i, j, 2), (i, j, 3))
                else:
                    d.union((i, j, 0), (i, j, 1))
                    d.union((i, j, 2), (i, j, 3))
                    d.union((i, j, 0), (i, j, 2))

        for i in range(N):
            for j in range(N):
                if (i + 1 < N):
                    d.union((i, j, 1), (i + 1, j, 3))
                if (j + 1 < N):
                    d.union((i, j, 2), (i, j + 1, 0))

        return d.count()


print(Solution().regionsBySlashes(["\\/\\ ", " /\\/", " \\/ ", "/ / "]))
